particle-iot/device-os
# Changes to support building combined Photon and P1 images for the recent releases

### Problem

We haven't done a manufacturing release for Photons and P1s since 0.5.5 and DeviceOS has changed quite significantly since then:

1. DCT functions do not fit into the bootloader and are dynamically imported from system parts
2. System flags are no longer stored in DCT
3. Tinker is no longer being built by default if no app is specified

### Solution

This PR along with the changes in https://github.com/particle-iot/photon-wiced/tree/feature/combined_fw introduces the following changes.

#### Changes the layout of the binaries in the combined image

```
Current layout:                            Previous layout:

+-----------------------------+ 0x08100000 +----------------------------------+
| mfg_test monolithic binary  |            |                                  |
|-----------------------------+ 0x080E5000 |                                  |
| Tinker (factory image)      |            | System part 2                    |
+-----------------------------+ 0x080E0000 |                                  |
|                             |            |                                  |
| mfg_test monolithic binary  | 0x080C0000 |----------------------------------|
|                             |            |
|                             |            |                                  |
+-----------------------------+ 0x080A0000 | System part 1                    |
|                             |            |                                  |
|                             |            |
| System part 2               | 0x08080000 +----------------------------------+
|                             |            |                                  |
|                             |            |                                  |
+-----------------------------+ 0x08060000 |                                  |
|                             |            |                                  |
|                             |            | mfg_test monolithic binary       |
| System part 1               |            |                                  |
|                             |            |                                  |
|                             |            |                                  |
+-----------------------------+ 0x08020000 +----------------------------------+
| EEPROM 2                    |            | Tinker (factory image)           |
+-----------------------------+ 0x08010000 +----------------------------------+
| EEPROM 1                    |            | EEPROM 1                         |
+-----------------------------+ 0x0800C000 +----------------------------------+
| DCT 2                       |            | DCT 2                            |
+-----------------------------+ 0x08008000 +----------------------------------+
| DCT 1                       |            | DCT 1                            |
+-----------------------------+ 0x08004000 +----------------------------------+
| Bootloader                  |            | Bootloader                       |
+-----------------------------+ 0x08000000 +----------------------------------+
```

Previously after the manufacturing app via the `wl` tool was done, the system parts were copied into their appropriate location by the bootloader and tinker was copied into the user part by the system parts.

Now, most of the things are in place and only the tinker is copied (from factory location) into its rightful place after the manufacturing app is done.

**NOTE:** tinker is placed into a gap in `mfg_test` binary, but is located at the correct factory image location (`0x080E0000`).

#### Introduces `MONO_MFG_FIRMWARE_AT_USER_PART` build flag

This flag causes the bootloader to be built with `MODULAR_FIRMWARE=y` and enables two things:
1. Jump into the user part location directly if there is a valid monolithic binary present @ 0x080a0000 (and is built to be run from 0x080a0000)
2. Certain system flags are read out from the DCT if condition 1 is satisfied

#### Build script/makefile

This PR removes obsolete targets from the makefile and introduces changes to layout the binaries in the order mentioned above.

### Steps to Test

1. set `WICED_SDK` env variable to point to the directory containing the photon-wiced repo, with the `feature/combined-fw` branch checked out
2. (optional) set `DEVICE_OS_DIR` env variable to point to the directory containing the device-os repo with the necessary release tag checked out
3. Run `hal/src/photon/make_combined.sh`, specifying the target PLATFORM_ID, e.g.: `./make_combined.sh PLATFORM_ID=6`
4. This will build the artefacts to `$(DEVICE_OS_DIR)/build/releases/release-$(DEVICE_OS_VERSION)-p$(PLATFORM_ID)` e.g. `$(DEVICE_OS_DIR)/build/releases/release-1.2.1-p6` for Photon
5. Run `openocd-flash` or `openocd-flash-stlink` target depending on whether `particle-ftdi` or `stlinkv2` debugger is used: `./make_combined.sh PLATFORM_ID=6 openocd-flash-stlink`
6. Connect a USB-to-UART adapter to TX/RX pins of your Photon and P1
7. The device should be booted into the manufacturing `mfg_test` application
8. Run the `wl.exe` tool from `$(DEVICE_OS_DIR)/build/releases/release-1.2.1-p6` (wine can be used): `wine wl.exe --serial 33 --interactive` (33 = COM33)
9. Execute `stm32id`, `setup_code`, `factory_reset`. All commands should execute successfully.
10. Reset the device, put into dfu. Run `particle keys doctor <deviceId>`. The device should reset and normally boot into tinker.
11. (This is an extra step added by @m-mcgowan that isn't part of the manufacturing process. It's to validate that when building the bootloader with `MODULAR_FIRMWARE=y` now the monolithic firmware still works) Flash monolithic firmware to the 0x8020000 and verify that it works.

### Example App

N/A

### References

- https://github.com/particle-iot/photon-wiced/tree/feature/combined_fw

---

### Changelog

- [internal] Changes to support building combined Photon and P1 images for the recent releases
